ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/api/impl/InMemoryLedgerReaderWriter.scala

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.api.impl
+
+import java.time.Clock
+import java.util.UUID
+
+import akka.NotUsed
+import akka.stream.scaladsl.Source
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueCommitting}
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlLogEntryId,
+  DamlStateKey,
+  DamlStateValue,
+  DamlSubmission
+}
+import com.daml.ledger.participant.state.kvutils.api.{LedgerReader, LedgerRecord, LedgerWriter}
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+import com.digitalasset.platform.akkastreams.dispatcher.Dispatcher
+import com.digitalasset.platform.akkastreams.dispatcher.SubSource.OneAfterAnother
+
+import scala.collection.JavaConverters._
+import scala.collection.{breakOut, mutable}
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.{ExecutionContext, Future}
+import com.google.protobuf.ByteString
+
+import scala.util.Random
+
+@SuppressWarnings(Array("org.wartremover.warts.ArrayEquals"))
+private[impl] case class LogEntry(entryId: DamlLogEntryId, payload: Array[Byte])
+
+private[impl] case class InMemoryState(
```

Comment:
case classes with mutable fields are problematic. Probably `InMemoryState` doesn't need to be a case class.

Reply:
Removed this anti-pattern.